product-catalog/app/productcatalog/controllers/ProductDetailPageController.java

```diff
@@ -0,0 +1,161 @@
+import common.prices.PriceFinder;
+import common.utils.PriceFormatter;
+import common.utils.Translator;
+import io.sphere.sdk.categories.Category;
+import io.sphere.sdk.products.ProductProjection;
+import io.sphere.sdk.products.ProductVariant;
+import play.Configuration;
+import play.libs.F;
+import play.mvc.Result;
+import productcatalog.models.ProductNotFoundException;
+import productcatalog.models.ProductVariantNotFoundException;
+import productcatalog.models.ShopShippingRate;
+import productcatalog.pages.*;
+import productcatalog.services.CategoryService;
+import productcatalog.services.ProductProjectionService;
+import productcatalog.services.ShippingMethodService;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static java.util.Locale.GERMAN;
+import static java.util.stream.Collectors.toList;
+
+@Singleton
+public class ProductDetailPageController extends SunriseController {
+
+    private final int numberOfSuggestions;
+    private final ProductProjectionService productService;
+    private final CategoryService categoryService;
+    private final ShippingMethodService shippingMethodService;
+
+    @Inject
+    public ProductDetailPageController(final Configuration configuration, final ControllerDependency controllerDependency, final ProductProjectionService productService, final CategoryService categoryService, final ShippingMethodService shippingMethodService) {
+        super(controllerDependency);
+        this.productService = productService;
+        this.categoryService = categoryService;
+        this.shippingMethodService = shippingMethodService;
+        this.numberOfSuggestions = configuration.getInt("pdp.productSuggestions.count");
+    }
+
+    private List<ProductThumbnailData> suggestionsToViewData(final List<ProductProjection> suggestions) {
+        return suggestions.stream().map((product) -> ProductThumbnailDataFactory.of(getTranslator(),  getPriceFinder(), getPriceFormatter()).create(product)).collect(toList());
+    }
+
+
+
+    public F.Promise<Result> pdp(final String slug, final String sku) {
+        //step 1 get all parameters from header/url
+        final Locale locale = GERMAN;//TODO get from URL
+
+        //step 2 create a chain/parallelize external calls
+        final F.Promise<ProductProjection> productProjectionPromise = fetchProduct(slug, locale);
+        final F.Promise<List<ProductProjection>> suggestionPromise = productProjectionPromise.flatMap(productProjection -> productService.getSuggestions(categoryService.getSiblingCategories(productProjection.getCategories()), numberOfSuggestions));
+        final F.Promise<CmsPage> cmsPagePromise = getPage("pdp");
+
+        //step 3 call method that does not need to call external systems
+        final F.Promise<Result> resultPromise = productProjectionPromise.flatMap(productProjection -> {
```

Comment:
the flatMap indendation has an interesting side effect, if you do more external calls, the code looks worse :)